### DISCUSSION TITLE
Ideas for developing three of Thinker's new features

### DISCUSSION THREAD

**Discussion by Michael2589** [induktio/thinker#25] [Ideas]
Three ideas for some of the new features

For new_base_names
Current behaviour:
Vanilla:
If it's the faction's first base, X. If it's a land base, Y1.
If it's a sea base, choose a name from the faction's #WATERBASES list of sea base names. If all of those have been used, choose a name from the faction's #BASES list of base names. If all of those have also been used, use the vanilla base name generator to create a name.
Thinker 2.8:
If it's the faction's first base, X. If it's a land base, Y2.
If it's a sea base, choose a name from the faction's #WATERBASES list of sea base names. If all of those have been used, use Thinker's base name generator to create a name.

How I'd like it to work:
If it's the faction's first base, X. If it's a land base, Y2.
If it's a sea base, choose a name from the faction's #WATERBASES list of sea base names. If all of those have been used, choose a name from the faction's #BASES list of base names. If all of those have also been used, use Thinker's base name generator to create a name.
Comment: This is both consistent with the original game's behaviour and means that you will get further into the game with flavorful base names.

Game setting: rare_supply_pods
You could mix in some monoliths among the nutrient/mineral/energy bonus resources. Monoliths are an interesting terrain feature, quite nice at the start of the game, and definitely not being transported from Earth on the Unity. I also think you could mix in some fungal towers, but that's just me. Remember, the vanilla supply pods aren't always good.

Game setting: revised_tech_cost=1
1. "Tech costs scale relative to the level of current research item, instead of the old version where costs increased according to the amount of techs a faction had."

I agree that this is the better choice when playing with directed research, but when you are playing with blind research the number of techs the faction has works better than the level of the tech being researched. (If I am playing with blind research, and I change my research priorities, then maybe the number of years expected to research the tech changes dramatically, which, besides just being weird, means that I can make a pretty good guess of what the tech is.)

Note that Progenitors always get directed research even when the game rules are set to blind research.

2. from details.md: "* Count every other faction with commlink to the current faction who has the tech, while infiltrated factions count twice. Discount 5% for every point in this value. Maximum allowed discount from this step is 30%. Only infiltration gained using probe teams counts for the purposes of extra discount, the Empath Guild or Planetary Governor status is ignored in this step."

Alternate choice: Factions that are either infiltrated or are in a Pact with the current faction count twice.

Comment: Historically, there are a lot of examples of allied countries sharing their technology.

**Comment by induktio** [comment]
Okay, let's see. :)

I'm not sure if I follow you with the base naming description. Thinker's new_base_names is supposed to use the same vanilla naming scheme until it runs out of names in the faction definition, and then generate new unique (generic) names for new bases. So not sure if there's a difference in the beginning.

About supply pods, it's certainly possible to add more monoliths on the maps. This can be done with the new world builder code. In fact, because the supply pod frequency is reduced, fewer monoliths are spawned on the map, so this could be adjusted closer to the original frequency. Maybe change like one fifth of the resource bonus tiles to monoliths.

Also it's true revised_tech_cost is more aimed for directed research games. I'm not really fan of the blind research mechanic as it randomizes gameplay way too much and also most multiplayer games leave it disabled as far as I know. Also I have a bias against the Progenitor factions because they don't fit very well thematically with the original game lore. So I'm not really trying to balance them as first priority. But yeah, it's true revised_tech_cost attempts to model some cross border "technology transfer" with those discounts. It always happens with some degree between nations, whether allied or not. Have to think about adjusting that bonus.

Also as advance notice, I'm planning to deprecate the old terranx_mod.exe and switch to a new exe launcher for the mod that does all the changes in-memory. So we would not need any patched terranx_mod.exe files anymore. After I deploy the changes, it would be good to gather some insight to make sure it works on all Windows versions. It should be available soon.

**Reply by Michael2589** [reply]
Each faction definition contains TWO lists of base names.  One is a list of generic names, and one is a list of names that are specifically for sea bases.  (Have a look at a faction file if this isn't making sense.)  The way the vanilla base naming algorithm works, is that:

if the base is a land base, it grabs a name from the generic name list, and if it's run out of those it names the base "Alpha Sector" or whatever. It ignores names from the sea base name list because most of them would look pretty strange on a land base.

If the base is a sea base, it looks for a name from the sea base name list, and if it's run out of those it looks for a name from the generic list, and if it's run out of those, too, it names the base "Alpha Sector" or whatever.

Your algorithm is doing the same thing as the vanilla algorithm for land bases, but for sea bases it's just checking the list of sea base names and, if there aren't any of those left, it gives the base an "Alpha Alpha Sector" style name, even if there are still unused names from the generic list.  I know this because I've founded a sea base and gotten an "Alpha Alpha Sector" style name, and then founded a land base and gotten one of the names from the faction's generic list.


Re: Supply pods:  That's what I was thinking about monoliths.

Re: directed research and blind research
I certainly wouldn't use blind research in a competitive multiplayer game, as better luck on randomly getting key techs early would almost certainly be more important in determining the winner than huge differences in player skill. However, I use it for single player games specifically because I like the gameplay being randomized.

The reference to Progenitors is because of something that I consider a vanilla bug rather than a balancing issue.  Think back to why you created the "tech_balance" option.  I don't want to use it in blind research games, for all the reasons that I'm playing with blind research, but I'd certainly use it in a directed research game.  But if I'm playing a blind research game with Progenitors, then some factions are using blind research and others are using directed research, and I can't set tech_balance=0 for some of them and tech_balance=1 for the Progenitors.  It's a minor and obscure issue though, and it's a problem that exists in the vanilla game, and I could fix it by tweaking alphax.txt if it was that important.

**Comment by induktio** [comment]
Starting from Dec 10 develop build, Thinker now contains a new launcher program thinker.exe which is used to start the mod.

- This requires the original GOG/non-GOG game binary terranx.exe in the same folder.
- Any patches are NOT written to terranx.exe, instead they are applied dynamically at runtime. Players can use terranx.exe to start the unmodified version of the game if necessary.
- Command line arguments should work as before. They are now given to thinker.exe.
- It is possible to use terranx.exe that is pre-patched with Scient's Patch v2.0 but any other versions of custom patches will probably fail at startup. Note that this is not necessary because the mod will apply Scient's Patch v2.0 changes at startup anyway.

If you have any issues with the launcher program, let me know. It could use more testing on different Windows versions, but probably it will be okay.

About the base naming scheme, I can see there's a difference with the way it works on sea bases vs. vanilla. However I can't decide now how to rewrite the code, because it uses a specific randomization scheme which should probably maintain compatibility with vanilla version. Have to think about it later.

**Reply by Michael2589** [reply]
I completely forgot to mention my experiences with the launcher program.

I'm running on Windows 10 Home version 21H1, Bitdefender antivirus, Intel Core i7. The first time I tried to start it, Windows recommended that I not run it because it's not a recognized program, and it was flaky the first few times I tried to start it (sometimes it would start, sometimes the program would just close before it did anything for no apparent reason), but since about the fifth time I started it, it's been fine. The issues may be because I was setting it up in a new folder to continue playing a game from an old folder, so Alpha Centauri.Ini was pointing at a game in a different folder, etc.

It actually works better than the previous version, because BitDefender was absolutely convinced that the older thinker.dll's were malicious software, but it's fine with the thinker.dll that comes with the launcher program.

**Reply by induktio** [reply]
Good to know it works well with Windows 10. I did not test that Windows version myself so that's important. I'm mostly finalizing some changes so we can bundle these in a new release.

There's been some issues with antivirus programs flagging the mod files with false positives. Antivirus heuristics become really nervous when they see some program patching binaries in memory, usually the detection is like HEUR/Malware or something generic. Not sure if there's a way to get completely rid of them other than maybe providing signed binaries, but that's a hassle. In any case there should be only very few false positives for thinker.exe now.

**Comment by induktio** [comment]
Version 2.9 is released now, but there's still some areas in the game balance that could be adjusted for next release. These could also involve additional alphax.txt edits.

Another player commented on reddit it's (too) easy to rush projects with crawlers. The AI doesn't currently rush projects with energy reserves or use crawlers for hurrying them either. This could be adjusted so that it would be harder to just grab every project later in the game, but it needs to also be balanced against how fun such an AI would be to play against. Thinker is not entirely about min-maxing for victory even though most of the mod is such.

* Full mineral disband ability could be removed from crawlers and make them disband the same amount than other units. When crawler upgrading is possible with energy reserves, this would be just too strong otherwise. Their purpose should be to gather resources, not be instantly disbanded after upgrade tricks.
* AI should not need to use crawlers to rush projects in such case, energy rushing would be sufficient. 
* AI could sometimes use their probe teams to subvert (lone) enemy units, but the implementation side binary stuff is not entirely reversed.
* AI currently doesn't evict neutral (non-hostile) probe teams from their territory. Maybe this should be implemented the same way the player can evict foreign probe teams from their territory?
* Config option limit_project_start could be removed. Note it doesn't affect the top 3 difficulty levels at default setting, but the option is kinda redundant for this kind of mod.

**Reply by Michael2589** [reply]
As observations, the vanilla AI will build crawlers to rush projects with (it's very bad at it, but it tries), and it is actually quite good/effective at subverting lone enemy units with its probe teams. You do NOT want to be fighting vanilla AI Morgan while running Knowledge unless you have the HSA!

The AI should DEFINITELY evict neutral probe teams from their territory the same as the player can. It didn't occur to me that it wouldn't.

Is it too easy to rush projects with crawlers? I'm not sure. If I've got a 60-base late game colony, I can certainly build projects in only a few turns with crawlers, but each turn takes much longer than turns earlier in the game, so I'm OK with it only taking a few turns.

I think it would be good for the AI to make some use of crawlers and energy credit rushing at higher difficulties (4+). Especially if you're playing at Transcend I think the player should expect the AI to use tactics.

However, it would emphatically not be fun for the player if the AI completes SP's suddenly and without warning. This can be solved without changing the game rules by restricting what the AI does. You could restrict how many crawlers an AI will cash at a SP each turn (e.g. 1/turn at difficulty 4, 2/turn at difficulty 5, 3/turn at difficulty 6, values need to be playtested to see how well they work, although it's OK for the AI to win the race if you're playing on Transcend; or perhaps the AI will only cash new crawlers at SP's and not cash in existing ones.) You could also restrict how much energy an AI will use to rush an SP each turn (e.g. the AI's net income at difficulty 4, twice the AI's net income that turn at difficulty 5, and 3x the AI's net income that turn at difficulty 6.)  Either of these rules will result in an AI SP proceeding at a sort of regular pace, letting the player guesstimate how long they have before the AI completes it.

The only alpha(x) modding strategy I've found for stopping the player (i.e. myself) from monopolizing SP's is to bunch up SP's at the beginning of the tech tree, so that there are several of them available at once. Then, because you can transfer production between SP's without penalty, the SP's are going to be split up between the factions trying to build them. I can try to bunch up SP's later in the tech tree, but if the player arrives at that point on the tech tree before the AI's do, the player can still have time to build all of them. I've found that increasing the average cost of SP's just makes the game more tedious, and trying to make end-game SP's extremely expensive is problematic because what's expensive for an end-game colony on an average-sized map is cheap for an end-game colony on a huge map.

Building SP's is one of the fun parts of the game, so you don't want to block the player from building lots of them in a successful play through.

I've wanted to have SP's that are faction-unique instead of world-unique since SMAC was published (and this was implemented in Civilization III as "minor wonders"), but the Supercollider is the only SMACX SP that I think clearly works well as a faction-unique wonder. (Faction-unique Supercolliders and world-unique Theory of Everything means that every faction can have a double-research base but only one faction can have a quadruple-research base.)  EDIT: The Ascetic Virtues actually works well as a faction-unique SP as well: I play with a pop limit of 16 if a base has a hab complex and I've disabled the actual Ascetic Virtues SP disabled, which plays out similarly to every faction having an Ascetic Virtues.

As you observe, you can reduce how easily players can rush projects with crawlers by removing the full mineral disband. You can also reduce how easily players can rush projects with energy credits by using the variable ec/mineral cost formula that's normally used for units instead of their normal 4ec/mineral cost.

The "upgrade tricks" are problematic, but if you're removing the full mineral discount from crawlers for this reason, you're doing it to stop players from doing it, since the AI won't do it unless you actively tell it how. I can see a use for such an option in multiplayer, but it's very low priority for single player since the player can simply refrain from using these "tricks" if they wish.

**Reply by induktio** [reply]
Hmm, when was the last time you saw vanilla AI build crawlers? Can't really remember any. It's been a while since I played against vanilla AI anyway. But if you spawn some crawlers on them, I remember they tend to move towards a base that is building some project, but they certainly don't know how to use the crawlers in a normal way.

Do you remember the AI issuing any withdrawal demands for probe teams on their territory? I'm not sure if it does. For normal units it should at least issue demands currently. I think that would be a simple way to make AI evict the probe teams, but it needs some more effort in deciphering the diplomacy variables. The funny thing is, no one has specifically requested this before, so it didn't occur to most players.

For the most part, diplomacy code is not modified by Thinker and the variables are not deciphered in detail. I've been doing some investigation on the code but what's annoying is that the net code is mixed in with it, and debugging legacy network code is quite painful. If I ever implement new diplomacy routines, they will probably have to exclude network multiplayer because it just doesn't make much sense to spend a lot of time debugging/testing it.

Removing the full mineral disband looks like a good choice now. In some areas I've already taken the path of simplifying some game rules.  So I'm not really in favor of implementing *both* crawler/energy rushing for AI factions. It just adds extra complexity, more situations for testing, and doesn't really add anything except maybe difficulty to the game.

If you open script.txt, there's a label called DONEPROJECT which can be used to warn player of energy rushing. If the AI rushes a project with energy reserves (based on some logic, not sure) this dialog could be issued, so the player would have maybe one turn to rush a project themselves, otherwise the project gets completed very quickly.

**Reply by Michael2589** [reply]
It's rare that the vanilla AI builds crawlers, but it does happen, and when I've seen it, it always looked like it had been trying to rush a SP. Rushing SP's with crawlers was a reskinned mechanic from Civilization II (and probably Civilization I), where you would rush Wonders with caravans - SP's were SMAC Wonders and crawlers are SMAC caravans - so Firaxis had coded that in the past. Crawlers lost the caravans' purpose of establishing trade routes and gained gathering resources from tiles and transferring resources between bases a new mechanics.

I don't remember the AI issuing withdrawal demands for probe teams on their territory.  The documentation (p. 130 of the SMAC manual), says under TREATY that "Both sides are expected to keep military units out of the other's territory (or at least remove them promptly when asked)."  If the AI has a colony pod on your territory and you demand they withdraw their units, one of the texts they can reply with is "As you must surely be aware, my $<> friend, I have no military units inside your borders." (Script.txt #NOWITHDRAW lines 3358-62). So the AI won't be demanding you withdraw probe teams under the treaty because they're not military units.

Mechanically, the way I make AI probe teams leave my territory during peacetime is that I attack them with one of my military units, and if it's peace and they're a probe team and they're on my territory I get a window with a menu in it, and one of the options in that menu is "Interrogate them, then return them unharmed".  The text is from #ENEMYPROBE on line 6492 of Script.txt. It doesn't arise from the diplomacy system.

I think I may have seen DONEPROJECT, although I don't think that it was because the AI was energy rushing, and I'm not sure whether it was in SMACX or Civilization I or II.  I found that if I responded to that warning by energy rushing my copy of the project so that it was 0 turns from completion, my project would be completed and not the AI's, but if I only rushed the project enough that it was 1 turn from completion, the AI's project would be the one completed.

**Reply by Michael2589** [reply]
Choosing to implement just energy rushing for AI factions makes a lot of sense.

When you "attack" a probe in peacetime with a military unit, you get three options: eliminating them, returning them, or leaving them alone. You could make the AI take into account the ai-fight ("willingness of the faction leader to use force to achieve goals) of the AI faction (from the faction.txt, e.g. Yang is 1, aggressive, Zakharov is 0, erratic, and Deirdre is -1, peaceful), and also the Mood of the leader, from Magnanimous to Seething, in deciding which of those three options they pick.

**Reply by induktio** [reply]
Yeah I know about the probe returning mechanic. ;) I've been investigating how to implement these and it looks like we'll be able to get all of these features. Just need to run a little bit more testing first.

* Secret project rushing
* AI using probes to subvert some units
* AI evicting probes from their territory
* Fix bug where foil probes get returned to landlocked bases

Many have attempted to fix that bug at different times, but I've implemented a solution that runs actual breadth-first pathfinding to find the closest base by travel distance. It's almost the same as if the player used normal unit pathfinding the find the closest faction base. The game engine never did this properly, so the issue kept repeating.

About the secret project rushing, it's possible I won't remove crawler mineral disband ability yet. Instead I might opt for a solution that makes the AI project rushing speed dependent on difficulty level. Specifically, on Transcend level when the AI is not ranked number one, it is able to rush projects such that they are left with 1 turn to completion. On all other difficulty levels, the rushing leaves something like 2-4 turns worth of production not completed, so the player can react to the DONEPROJECT popup. I've implemented it such that you require commlink to the specific faction and not have sunspots active to receive those messages.

**Reply by Michael2589** [reply]
Removing the crawler mineral disband ability is going to be an option, like everything else important? Not telling the AI how to do it is one thing, but it's a big change from the vanilla rules. (I'm just reacting to your phrasing in the last post: it's sometimes hard to tell what the other person is assuming when one is communicating completely over the Internet like this.)

I've thought before about making crawlers like other units - costing maintenance, and you get back half your minerals if you scrap them, like transports.  I think that works well conceptually, and it would also be somewhat better game balance. At the moment, you can get a good rate of return on crawlers by simply crawling forest tiles, but if they cost maintenance it would reduce the profitability of crawling forest tiles by half while not impacting their profitability on more specialized tiles much.

I'm not sure what the implications for armored crawlers in combat are.  At the moment, if you defeat a military unit in a stack, the other military units take "collateral damage" unless they are in a base or a bunker.  If you defeat the last military unit in a stack, civilian units such as probe teams, crawlers(?), and mind worms(!) are destroyed, even if the stack is in a base.  I think that if armored crawlers cost maintenance, then they should be treated as armored transports for purposes of collateral damage, although I'm not sure whether that's different from how they're currently treated.  (I'd definitely like to see mind worms treated as military units for purposes of collateral damage, and that would make some of the fungal blooms more of a threat than the lucrative events they are now.)

At the moment sunspots are pretty arbitrary in their effects - they shut down diplomacy while commerce and spying continues.  I can't imagine how that would work - if your merchants can trade with each other, they should be able to carry messages.  If you are the Governor, or have the Empath Guild, or have simply infiltrated the other faction, you're going to be able to see that they've nearly completed the SP by accessing the F4 Bases report or examining the base building the SP even if you're not getting the popup message.  It would be interesting to have this effect as part of a comprehensive review of the effects of sunspots.  At the moment, the effects of sunspots are arbitrary and then you're adding this one piece of logical consistency ;) which seems odd amidst all of the arbitrariness.

I noticed, in my last game, an issue with the Social Engineering AI and bases being mind-controlled which will become worse if the AI's start subverting each others' units.  The AI's favourite SE combination in that game was Democracy-Planned-Knowledge, which I was using myself, and it's a good combination if you're at peace or if you have the HSA. However, the Drones were in a Vendetta with Morgan, and the fact that the Drones were running Knowledge meant that every time the Drones captured a base from Morgan, Morgan promptly mind-controlled it back.  It would be good if you could increase the weight of the Probe stat to the SE AI when the faction is in a serious war, unless they have the HSA (and possibly even then, if their enemies have Algorithmic Enhancement).

Here's a possible addition to the list of vanilla bugs:
The game doesn't keep track of base culture very well.
If you capture a base from an enemy, if it's the base is their culture then it remains their culture, and if it is your culture then it remains your culture and the damage done during the conquest is less, but if it is a third party's culture, i.e. a base that the enemy faction captured from someone else, then it changes to the culture of the faction you captured it from.  I think it should remain the third party's culture.
If you give a base to someone in diplomacy, then it becomes their culture regardless of which culture it was. (This is handy when you're fighting an enemy and turning over the bases to a subject faction!)  I think that the only bases that should change culture in this situation are bases that are the culture of the faction turning them over:  bases of third party cultures should remain the third party culture.
I'm not sure whether that counts as a bug or an unintuitive feature.

**Reply by induktio** [reply]
Okay, pushed a new build now. This adds several important features that should also increase the challenge level.

* Secret project rushing (only those projects that match AI priority weights the most)
* AI using probes to subvert some units (this especially might need more testing)
* AI evicting probes from their territory
* Fix bug where foil probes get returned to landlocked bases
* Crawler issue is dealt with by changing the upgrade cost so that it offers no additional benefit for just upgrading crawlers for secret project rushing

In addition, this build might be the first time I'm directly modifying diplomacy dialogue in the game. :) Old AIs had an issue where they granted loans with very unbeneficial conditions for them, so now Thinker calculates whether the loan is granted and how much is offered. Dialogue path "urgent need of energy credits > schedule of loan payments" activates this feature.

" If you defeat the last military unit in a stack, civilian units such as probe teams, crawlers(?), and mind worms(!) are destroyed, even if the stack is in a base." <- This can't be true, there shouldn't be stack damage inside bases? Outside bases, probe teams and (planet-owned) mind worms are destroyed as a stack though.

**Reply by Michael2589** [reply]
It's not exactly stack damage. If, for example, the AI has 3 military units and 2 probe teams inside a base, and you defeat the first military unit, there's no damage to the other units. Defeat the second military unit, same thing. Defeat the third, and the probe teams die at as well, and if you're looking in the right place on the screen you'll see the message "3 units were destroyed" briefly displayed.

I know it's weird. (That's why I mentioned it.) But I've been paying attention to it, because it's the difference between being able to use armoured probe teams as maintenance-free base garrisons or not.

I can't remember now whether this is how it works in vanilla.

**Reply by induktio** [reply]
Hmm. Thinker does not change those stack damage rules currently. I think that probe team stack destruction rule should be removed at least when they are inside bases. It doesn't make much sense. Seems almost like an oversight in the game. If I remember right, it's processed somewhere in the long and convoluted battle_fight function. Outside the bases it could maybe apply some rule that if an armored probe is defending the tile, stack damage destruction does not apply or something. But I'm not sure how easily those can be patched in.

**Reply by Michael2589** [reply]
I think it's completely reasonable as a feature.  That is, being able to use maintenance-free units interchangeably with military units that cost maintenance as defensive troops has a lot of implications. Especially since the armored probe team is actually an offensive threat, since it can subvert units, and can move while ignoring zones of control, and all of that despite being cheap... they don't need to be made stronger.

**Reply by induktio** [reply]
On the defensive side, that feature shows probes are not that strong. Supposedly if a base contains 5 armored probes and one garrison, they all get wiped out after the garrison loses which seems nonsensical. Outside bases (or bunkers), you could make that argument they need not be made stronger, but that's a stretch to say the current version is a good mechanic for base tiles. At one point I thought about implementing this earlier, but decided to skip it back then when coincidentally I implemented some mechanics that force the AI to disperse its troops into multiple tiles. That doesn't help with stack destruction on base tiles though. Might be a good idea to have some kind of a config option for that.

**Reply by induktio** [reply]
On another note, just looked throught the base production code and confirmed the game engine has extra AI cheats for facility maintenance. The AI pays only 1/3 maintenance cost for facilities on transcend and 2/3 on thinker level. On other levels the costs are same as for player factions. The numbers that the game shows on infiltration screens do NOT account for AI bonuses on maintenance, the bonus is just quietly calculated during turn processing. These are probably not documented at all, at least I don't remember seeing them mentioned. Seeing how the AI accumulates credits on transcend level can look strange. The bonus on that level is a little high and might need some rebalancing.

**Comment by Michael2589** [comment]
Obviously, they're not that strong as base defenders (i.e. probes can't be used as base defenders against a military assault). But they shouldn't be useful against military assault. That's not their purpose.

Why should I build base defenders that cost maintenance if I can build base defenders that don't?

As for why it makes sense, the fiction is that they are probe TEAMS, not probe regiments or probe divisions. The number of people in them is considered trivial compared to the number of people in the military units.

It would be reasonable for them to escape a falling base (e.g. use the function that sends them home after completing a mission to send them to a different base owned by their faction, though what if this was the faction's last base but the faction is still alive due to colony pods?) It is not reasonable for them to prevent enemy military units from occupying an otherwise undefended base.

Something that would be nice would be to increase the cost of mind controlling a base depending on the number of military units in it. At the moment, if you have two military units in a square you can't subvert them, but if you have 20 military units in a size 2 base you can get the base and the military units for the cost of mind controlling an empty size 2 base.

I remember the 1/3, 2/3 maintenance on AI base facilities, although I can't find any mention of it.  Supposedly it was bugged in SMAC, so that the AI paid triple maintenance on their base facilities at transcend difficulty in SMAC, and this was fixed in SMAX.

I don't think there's a need to change this.  It makes sense in that the AI isn't checking that it has enough revenue to pay for its maintenance, and on higher difficulty levels it'll randomly build more of everything, so more base facilities, so more maintenance.

Can you check whether there is actually a research penalty associated with punishment spheres?  As far as I can tell, looking at the various screens in game, that isn't implemented, but it's entirely possible that it is and it's just not being reported on the report screens.

**Reply by induktio** [reply]
Not all of the effects are shown in the GUI obviously, if we dig into base_research this piece of code is run there:

> v1 = current_base_id;
> v2 = current_base_ptr->labs_total;
> bitmask(FAC_PUNISHMENT_SPHERE, &v8, &v7);
> if ( (unsigned __int8)v7 & bases.facilities_built[308 * v1 + v8] ) {
>    v2 /= 2;
> }

Some people would like to change the mind control mechanics, but opinions differ, so go figure. Actually when you noted about the probe ratings and mind control, I implemented a small priority change that keeps track when the last base was lost to mind control. If it was recently, social_ai should put additional weight for probe rating. 

This was actually really simple to implement because there was already code related to it. This and some other tweaks for loan diplomacy dialog are implemented in Dec 22 build. The AI has new dialog now when they strongly reject a proposed loan.

**Reply by Michael2589** [reply]
So the research penalty for punishment spheres was implemented, but, whether or not it is buggy, I have no idea.  I'm visiting family right now but when I get home I'll look into setting up a test for it.

Thanks for the change to SE if they lose bases to probe teams :).

If that's the same terrible loan offer that the AI offers the player, good riddance. Now if you could only stop the AI from asking the player for loans that would be perfect.

I was trying to figure out situations in which a loan could be a reasonable idea. In faction.txt, one of the abilities you can give to a faction is interest on their energy balance. So if a faction had the ability to collect, for example, 10% interest on their energy balance, it would make sense for them to suggest, to factions they don't dislike, that the other faction give them a loan for (in that example) 5% interest,  and then the principle would get repaid as a block after, for example, 10 turns.  The faction acting as banker would then keep the excess interest.  But since no factions actually have this ability the situation won't arise.  And I can't think of any other situations where giving loans actually makes sense.

While you're adding features to the sunspots event to flesh it out, you could give a configuration option in thinker.ini that sunspots would also cut commerce in half. If things are so bad that faction leaders can't communicate, the presumably commerce would be adversely affected as well.  I do like the idea of leaders using the cover of sunspots to nerve staple troublesome bases!

**Reply by induktio** [reply]
Where do you get the idea sunspots work somehow differently in the mod? It should work similarly to vanilla game, you can nerve staple bases without commerce penalty when sunspots are active. Also that DONEPROJECT has a similar sunspots check before it is shown.

So right now the AI affinity for loan granting depends heavily on treaty status, diplomatic friction, integrity blemishes and the relative size of the factions (they are more willing to loan to smaller factions) plus a couple of smaller factors. The diplomacy code is quite convoluted and mixed in with the GUI code, so I'm not going to attempt to replace large fractions of it, but some smaller tweaks on some specific areas could be done.

**Reply by Michael2589** [reply]
The reference to nerve stapling bases without commerce penalty when sunspots are active was a reference to the change in AI behaviour in patch 2.9.  It wasn't intended to imply that you'd changed the game rules.

**Comment by induktio** [comment]
Previously Thinker Mod has not modified the tech tree but I've been developing some interesting concepts recently. [The files can be downloaded here.](https://www.dropbox.com/sh/qsps5bhz8v020o9/AAAkyzALX76aWAOc363a7mgpa/resources) There were some very specific design ideas that I decided to implement. I have not decided if I should make this new tree the default in the mod, probably needs more comments from players.

* Provide some in-between option for smacx and smac-in-smacx mod with the amount of game content.
* Rearrange all weapon and armor techs such that they are discovered in a linearized sequence, one for each type, no exceptions. Each type has a multiple tech delay before a better one can be discovered.
* Remove all expansion weapons/armors. They just clutter the tree with redundant options.
* Remove all expansion techs (and by addition, most of all references to Progenitors).
* Keep all expansion abilities/facilities/secret projects but move them to appropriate vanilla smac techs.
* Add Global Energy Theory and Inertial Damping techs that were cut from the original game. These are good techs for some of the smacx items.
* As it can be seen from the graph, a special property of this new tech tree is that techs from level n have only prereqs from level n-1. First 7 levels also have exactly 7 techs each.
* This approach forces to prioritize early-to-midgame techs. For example Neural Grafting arrives later than in the base game but I think it's okay. For early game one has to now choose carefully which special ability to put on the units.
* Hovertanks and gravships arrive slightly earlier than in the original game. Graviton Theory could even be earlier but not really sure.
* Some important reorderings: satellites are now discovered in a sequence Fusion power -> Orbital Spaceflight -> Super Tensile Solids -> Advanced Spaceflight. Biomachinery is moved later because it contains an OP project, and so forth.

**Reply by JoGr223** [reply]
"Rearrange all weapon and armor techs" I don't think it solves any actual problem in the game ?

"Remove all expansion techs" I don't like it, it looks like solution in search of a problem; I like that for example PEG SP has its own tech (Adaptive Economics) as well as it gives thermocline transducer

"Keep all expansion abilities/facilities/secret projects but move them to appropriate vanilla smac techs." that's the key thing and you are not showing how you are going to do this

"Hovertanks and gravships arrive slightly earlier than in the original game. Graviton Theory could even be earlier but not really sure." you are not stating what's the point of all these changes; the hovertank is actually what i've been thinkig about too (and i think one of the mods might already have that); but the problem here is with Doctrine Air Power - it would be better if hovertanks were before planes

I think you are making deep changes without any reason - each of this proposal should have separate topic :) it feels rushed and too global; better to make small directed changes where each solves something; I originally thought Thinker mod is only about AI, then I learnt it makes slight rebalance like slower research and no attack bonus from better reactors; I'd prefer to make only necessary changes; readjusting whole tech tree globally does not look like something that's needed

**Reply by induktio** [reply]
Let's just clarify a couple of things. You can click on the dropbox link and it should contain the exact same alphax.txt that was used to draw the graph. You can also install it in the game and look at datalinks, so it's not like the implementation is "hidden" in any way. Planetary Energy Grid would get moved to Global Energy Theory and it's quite a fitting tech for the project.

Also let's be clear about the fact that the mod will not require having any specific tech tree while playing. Already the mod ships with smac-in-smacx as an alternative, so the new tech tree might get added as an additional choice. If some people like to have just the default settings, there's nothing preventing it. Of course then there are the binary patches, but thinker.ini file is for them.

There are many players who may not particularly like the expansion stuff but would like to keep some of them. And some of them just play smac-in-smacx. Even though I didn't describe all of the rationale there, I don't really see how you can say they're just unnecessary changes. :) For example removing the expansion armors/weapons has already been done in Will To Power, because those items basically duplicate existing stuff and add very little to the game.

Also weapon/armor reordering serves a specific purpose: it makes each of them have a specific timeframe where they can be used. In the vanilla game some of them nearly always get skipped over because a better one is available at the same time. Sure, there are many changes, but also on other hand, I've tried to keep the previous ordering for many techs where possible. So here we can discuss it. :) That hovertank/needlejet reordering could be possible I suppose, that's something to think about.
